args_test.go

```diff
@@ -5,231 +5,208 @@ import (
 	"testing"
 )
 
-func TestNoArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: NoArgs, Run: emptyRun}
+func newCommand(args PositionalArgs, withValid bool) *Command {
+	c := &Command{
+		Use:  "c",
+		Args: args,
+		Run:  emptyRun,
+	}
+	if withValid {
+		c.ValidArgs = []string{"one", "two", "three"}
+	}
+	return c
+}
 
-	output, err := executeCommand(c)
+func expectSuccess(output string, err error, t *testing.T) {
 	if output != "" {
-		t.Errorf("Unexpected string: %v", output)
+		t.Errorf("Unexpected output: %v", output)
 	}
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
 }
 
-func TestNoArgsWithArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: NoArgs, Run: emptyRun}
-
-	_, err := executeCommand(c, "illegal")
+func expectError(err error, t *testing.T, ex string) {
```

Comment:
Not a fan of this helper, the keys in the map are basically magic values outside this function.

Reply:
The keys in the map correspond to the validator that we expect to get an error from:

- `validateValidArgs`: `valid`
- `NoArgs`: `no`
- `ArbitraryArgs`: does not produce errors
- `MinimumNArgs`: `min`
- `MaximumNArgs`: `max`
- `ExactArgs`: `exact`
- `RangeArgs`: `range`

I can make it more explicit by just stripping `Args`, if you want:

- `validateValidArgs`: `validateValid`
- `NoArgs`: `No`
- `ArbitraryArgs`: does not produce errors
- `MinimumNArgs`: `MinimumN`
- `MaximumNArgs`: `MaximumN`
- `ExactArgs`: `Exact`
- `RangeArgs`: `Range`

Reply:
Well first off, comparing error text is always a bad idea even in tests. It's just brittle.
But that is pretty much what the code was already doing anyways.  I get that you wanted to reduce duplication, but in this case I think some error constants would be more clear and you can nix, the helper function.

Reply:
The main objective was not to remove code duplication, but to make two characteristics explicit and easy to spot:

- Whether the test is expected to success or fail.
- Whether multiple test cases are expected to return exactly the same result.

For example:

``` go
func TestMinimumNArgs_WithValid_WithInvalidArgs(t *testing.T) {
...
	expectError(e, t, "valid")
...
func TestMinimumNArgs_WithLessArgs(t *testing.T) {
...
	expectError(e, t, "min")
...
func TestMinimumNArgs_WithValid_WithLessArgs(t *testing.T) {
...
	expectError(e, t, "min")
...
func TestMinimumNArgs_WithValid_WithLessArgsWithInvalidArgs(t *testing.T) {
...
	expectError(e, t, "valid")
```


``` go
func TestMinimumNArgs_WithValid_WithInvalidArgs(t *testing.T) {
...
	expectError(e, t, `invalid argument "a" for "c"`)
...
func TestMinimumNArgs_WithLessArgs(t *testing.T) {
...
	expectError(e, t, "requires at least 2 arg(s), only received 1")
...
func TestMinimumNArgs_WithValid_WithLessArgs(t *testing.T) {
...
	expectError(e, t, "requires at least 2 arg(s), only received 1")
...
func TestMinimumNArgs_WithValid_WithLessArgsWithInvalidArgs(t *testing.T) {
...
	expectError(e, t, `invalid argument "a" for "c"`)
```

Reply:
> * Whether the test is expected to success or fail.
> * Whether multiple test cases are expected to return exactly the same result.

The first can be accomplished just by checking for the presence of an error value then.
The second is trickier as currently the only way of telling different validation failures apart is by the error text.  This wasn't an issue until now because the ValidateArgs function only ran 1 validator.

Reply:
> You could approach the checks a few ways,
>     1. use constants for the error text

I would really like to do this. Not just because of this PR, but specially because of #842. The idea there is to slighly rework the behaviour when execution errors are produced either because the command is not runnable, because it does not accept args, because args are invalid or because the arg is neither a valid arg nor a valid subcommand.

- `invalid argument "a" for "c"`
- `command "c" is not runnable`
- `command "c" is not runnable; please provide a subcmd`
- `unknown command "one" for "c"`
- `"one" rejected; "c" does not accept args`
- `requires at least 2 arg(s), only received 1`
- `accepts at most 2 arg(s), received 3`
- `accepts 2 arg(s), received 3`
- `accepts between 2 and 4 arg(s), received 1`

Note that the management of the first five error strings above is directly related to deciding when `[args]` should be shown in the usage/help.

However, the error strings above are not constant. `"a"`, `"c"`, `"one"`, `2`, `1`, `3`, `4`... all of those values are generated when the error is produced. Should I define 'error constants' as `invalid argument %s for %s` and:

- Use the 'constant string template' to generate the error.
- When I need to identify it, execute the template with the current `cmd` and compare it with the err.

How can I ensure that the `cmd` when the error was generated is the same as the `cmd` in the function where it is being evaluated?

>     2. use sentinel errors.

I don't know what are 'sentinel errors'. Isn't it the same as the procedure described above?

https://dave.cheney.net/2016/04/07/constant-errors

>     3. Rely on error behavior (see github.com/pkg/errors)

This seems to be exactly what we need.

> Latter option is best but entails a larger change-set, constants or sentinel errors probably the easiest.

I'll try to add it to #842. But I think it should not be a requirement for this PR.

> Whatever solution, we shouldn't be comparing error text in the tests.

Agree. But we are already doing it everywhere:

- https://github.com/spf13/cobra/blob/master/args_test.go
- https://github.com/spf13/cobra/blob/master/bash_completions_test.go
- https://github.com/spf13/cobra/blob/master/cobra_test.go
- https://github.com/spf13/cobra/blob/master/command_test.go
- https://github.com/spf13/cobra/blob/master/zsh_completions_test.go

I think it is out of the scope of this PR to fix it. Moreover, I think that 'errStrings' is a step forward towards implementing 3, since ot allows the list of expected errors to be listed somewhere.

Reply:
Agreed that it overreaches the scope. It should probably be a standalone PR.
Going to go through some of this again this evening and then ill give it my 👍 . 

Check my comment below about organization of the test structure.